invenio_rdm_records/records/systemfields/access/grants.py

```diff
@@ -118,9 +118,9 @@ def to_token(self):
         # this ensures that we can always separate the values again
         # (b/c the base64 alphabet doesn't contain the dot)
         return "{}.{}.{}".format(
-            b64encode(str(self.subject_type).encode()).decode(),
-            b64encode(str(self.subject_id).encode()).decode(),
-            b64encode(str(self.permission_level).encode()).decode(),
+            b64encode(str(self.subject_type).encode(), b"-_").decode(),
+            b64encode(str(self.subject_id).encode(), b"-_").decode(),
+            b64encode(str(self.permission_level).encode(), b"-_").decode(),
```

Comment:
:thinking: does that mean a query for permission tokens need to know to b64encode its arguments beforehand to sidestep strange issues? ... I am probably naive, but if we don't want dots, why don't we have the code enforce that and raise an exception if a dot is present in those arguments? Those arguments are all internal code anyway (I don't see anything wrong with that being the way we roll).

Reply:
we can of course change the encoding to something that's easier to use from a programming perspective, i just selected base64 because i thought it offered some nice benefits:
* with the alternate chars "-" and "_", we always get a string token that's URL-safe
* we're completely independent on the actual values of `subject_type`, `subject_id` and `permission_level` -- they can include any characters
* b64 results never contain dots, so we can be sure that dot-separation will always work

if we know an *entire* token, we can always do something like `Grant.from_dict({"subject": "user", "id": "1", "level": "view"}).to_token()` and not worry about the concrete way that the tokens are generated (in the back-end at least)

however, if we want to generate partial tokens (e.g. from `{"subject": "user"}`), the current approach probably won't work as well...
in that case, we might also want to change the ES mapping however, since i think that the `keyword` type doesn't allow for substring matches:
https://github.com/inveniosoftware/invenio-rdm-records/blob/master/invenio_rdm_records/records/mappings/v7/rdmrecords/records/record-v1.0.0.json#L54

Reply:
I am thinking about the Generator query filters: those generators would be creating the tokens. Most of those are in invenio-records-permissions, but records-permissions shouldn't have a dependency on invenio-rdm-records (otherwise circular). So  if we are relying on `Grant.from_dict({})`, we should put those generators in invenio-rdm-records... which means all(?) generators should be placed here (not quite what we wanted :smile_cat: ) If we are not relying on `Grant.from_dict({})`, then we need to be careful to copy this encoding in that module and keep them in sync... 

In any case, we will be diving deeper into those questions soon, so we can assess that separately.

Reply:
ahhh, i see!
yes, that's kind of an issue :confused:

Reply:
Most of the generators will be in Invenio-RDM-Records as they are specific to the RDM data model.